### ISSUE TITLE
Example in "Compile and run TFLITE with SystemC"

### ISSUE THREAD

**Issue by TommyR06** [tflite-soc/tflite-soc#9]
Hi!
Is it correct the example in the README about SystemC with TFlite?
I am talking about this part:
>#2D Systolic array accelerator with TFLite benchmarking tools
> bazel build --jobs 1 //tensorflow/lite/examples/systemc:hello_channel # SeeBugs
> bazel build //tensorflow/lite/tools/benchmark:benchmark_model --cxxopt=-DTOGGLE_TFLITE_SOC=1
> bazel-bin/tensorflow/lite/tools/benchmark/benchmark_model \
>     --use_gpu=false --num_threads=1 --enable_op_profiling=true \
>     --graph=../tensorflow-models/mobilenet-v1/mobilenet_v1_1.0_224_quant.tflite \
>     --num_runs=1 

I cannot understand the reason of building the hello_channel example here.
Assuming that the correct program is the systolic_run example, I still have some doubt about the TOGGLE_TFLITE_SOC definition. 
As far as I can see, it just turns on some PrintMatrix functions. 
Where is the integration between SystemC and TFlite?
I am clearly misunderstanding something .. any help would be highly appreciated!

**Comment by linbaiwpi** [comment]
> Hi! Is it correct the example in the README about SystemC with TFlite? I am talking about this part:
> 
> > #2D Systolic array accelerator with TFLite benchmarking tools
> > bazel build --jobs 1 //tensorflow/lite/examples/systemc:hello_channel # SeeBugs
> > bazel build //tensorflow/lite/tools/benchmark:benchmark_model --cxxopt=-DTOGGLE_TFLITE_SOC=1
> > bazel-bin/tensorflow/lite/tools/benchmark/benchmark_model 
> > --use_gpu=false --num_threads=1 --enable_op_profiling=true 
> > --graph=../tensorflow-models/mobilenet-v1/mobilenet_v1_1.0_224_quant.tflite 
> > --num_runs=1
> 
> I cannot understand the reason of building the hello_channel example here. Assuming that the correct program is the systolic_run example, I still have some doubt about the TOGGLE_TFLITE_SOC definition. As far as I can see, it just turns on some PrintMatrix functions. Where is the integration between SystemC and TFlite? I am clearly misunderstanding something .. any help would be highly appreciated!

Hi Tommy,

I think the correct first command is "systolic_run" rather than "hello_channel". But it seems utils will anyway be compiled. Nothing will be changed. Concerning to the print functions. My opinion is, it did not change the GEMM function of TFLITE inference procedure but shows how to integrate the systemC module into this inference flow.

If you would love to integrate your own functions. You have to replace the corresponding functions with your systemC functions (like GEMM or softmax). I have tried to call the `hello_world` module and inserted into `SoftmaxEval` function. After re-compiling and running, I can see the "Hello World!" printed out.

```cpp
TfLiteStatus SoftmaxEval(TfLiteContext* context, TfLiteNode* node) {
  auto* params = reinterpret_cast<TfLiteSoftmaxParams*>(node->builtin_data);
  SoftmaxOpData* data = reinterpret_cast<SoftmaxOpData*>(node->user_data);

  const TfLiteTensor* input = GetInput(context, node, 0);
  TfLiteTensor* output = GetOutput(context, node, 0);
// MY CODE BEGIN
#ifdef TOGGLE_TFLITE_SOC
  tflite_soc::hello_world hello("HELLO");
  hello.say_hello();
#endif
// MY CODE END
  switch (input->type) {
    case kTfLiteFloat32: {
      return SoftmaxFloat(context, input, output, params);
    }
```

Hope this helps.

BR,
Lin
